# Week 2 - Day 3: Linear Algebra for Machine Learning

In this notebook, I applied the basic linear algebra concepts used in machine learning. I represented car sensor readings as vectors and matrices, calculated predictions using the dot product, and used matrix multiplication to process multiple samples at once.

I also explored vector magnitude, matrix transpose, determinants, inverses, eigenvalues, and common shape mismatch errors.




## 1. Importing NumPy

NumPy provides efficient tools for working with vectors, matrices, and linear algebra operations.

The print precision is set to two decimal places to make numerical results easier to read.## 1. Importing NumPy

NumPy provides efficient tools for working with vectors, matrices, and linear algebra operations.

The print precision is set to two decimal places to make numerical results easier to read.

In [1]:
import numpy as np
np.set_printoptions(precision=2)

## 2. Representing Car Samples as Vectors

Each car is represented as a one-dimensional NumPy array.

The three values in every vector represent:

1. Engine capacity in cubic centimeters
2. Vehicle speed in kilometers per hour
3. Engine temperature in degrees Celsius

Each car vector therefore contains three features.

In [2]:

car1 = np.array([2200 , 60 , 88])
car2 = np.array([3100 , 95 , 93])
car3 = np.array([1800 , 40 , 84])
print(car1)
print(car2)
print(car3)


[2200   60   88]
[3100   95   93]
[1800   40   84]


Each vector represents one sample:

- `car1 = [2200, 60, 88]`
- `car2 = [3100, 95, 93]`
- `car3 = [1800, 40, 84]`

In machine learning, one sample is commonly represented as a vector whose elements are the input features.

## 3. Creating the Dataset Matrix

The three car vectors are combined into one matrix named `cars`.

In a machine learning dataset:

- Each row represents one sample.
- Each column represents one feature.

The expected shape is:

$$(\text{number of samples}, \text{number of features}) = (3, 3)$$


In [3]:
cars = np.array([car1 , car2 , car3])
print(cars)
print(cars.shape)

[[2200   60   88]
 [3100   95   93]
 [1800   40   84]]
(3, 3)


The shape `(3, 3)` means that the dataset contains:

- 3 car samples
- 3 features for each car

This row-by-column organization is the standard structure used by most machine learning libraries.

## 4. Defining the Model Weights

A weight vector assigns a level of influence to each input feature.

The weight vector contains three values because every car sample contains three features.

$$w = [0.003,\ 0.5,\ 0.2]$$

The first weight is applied to engine capacity, the second to speed, and the third to engine temperature.

In [4]:
weights = np.array([0.003 , 0.5 , 0.2])

## 5. Computing a Prediction Using the Dot Product

A simple linear model calculates its prediction using the dot product between a feature vector and a weight vector.

For the first car:

$$x_1=[2200,\ 60,\ 88]$$

$$w=[0.003,\ 0.5,\ 0.2]$$

The dot product is calculated manually as:

$$x_1 \cdot w
=
(2200 \times 0.003)
+
(60 \times 0.5)
+
(88 \times 0.2)$$

$$=6.6+30+17.6$$

$$=54.2$$

The dot product combines the three input features into one numerical prediction. This is the central calculation performed by linear models and artificial neurons before adding a bias or applying an activation function.

In [5]:
prediction_car1 = np.dot(weights, car1)
prediction_car2 = np.dot(weights, car2)
prediction_car3 = np.dot(weights, car3)

### Verifying the Dot Product with NumPy

The manual result can be verified using `np.dot()`.

The same operation is repeated for all three cars. Because both inputs are one-dimensional vectors with three elements, the result of each dot product is a single scalar prediction.

In [6]:
print("Car 1 prediction:", prediction_car1)
print("Car 2 prediction:", prediction_car2)
print("Car 3 prediction:", prediction_car3)

Car 1 prediction: 54.2
Car 2 prediction: 75.4
Car 3 prediction: 42.2


The calculated predictions are:

- Car 1 prediction: `54.2`
- Car 2 prediction: `75.4`
- Car 3 prediction: `42.2`

A larger result means that the weighted combination of the car features is larger. The exact meaning of the score depends on what the model was trained to predict.

## 6. Predicting All Samples with Matrix Multiplication

Instead of calculating one dot product at a time, matrix multiplication can calculate predictions for all samples in one operation.

The dataset matrix has the shape:
$$cars:(3,3)$$

The weight vector has the shape:

$$weights:(3,)$$

The three columns in the dataset match the three elements in the weight vector. Therefore, the multiplication is valid.

$$(3,3)@(3,)\rightarrow(3,)$$
The result is a vector containing one prediction for each of the three cars.

In [7]:
predictions = cars @ weights

Matrix multiplication applies the same weight vector to every row in the dataset.

The result is equivalent to calculating:

$$
\begin{bmatrix}
car_1 \cdot w\\
car_2 \cdot w\\
car_3 \cdot w
\end{bmatrix}
$$

Using matrix multiplication is more efficient and scalable than calculating every prediction separately.

In [8]:
print("All predictions:", predictions)
print("Predictions shape:", predictions.shape)

All predictions: [54.2 75.4 42.2]
Predictions shape: (3,)


## 7. Matrix Transpose

The transpose operation exchanges the rows and columns of a matrix.

Originally:

- Rows represent car samples.
- Columns represent car features.

After transposing the matrix:

- Rows represent features.
- Columns represent samples.

The transpose is frequently used in machine learning calculations, especially when computing covariance matrices, gradients, and linear regression equations.

In [9]:
print(cars.T)

[[2200 3100 1800]
 [  60   95   40]
 [  88   93   84]]


## 8. Pairwise Dot Products

Multiplying the car matrix by its transpose produces a matrix containing dot products between every pair of car samples.

$$cars:(3,3)$$

$$cars^T:(3,3)$$

Therefore:

$$cars@cars^T:(3,3)$$

Each element in the resulting matrix measures the raw numerical similarity between two car vectors.

In [10]:
print(cars.T)
print(cars @ cars.T)

[[2200 3100 1800]
 [  60   95   40]
 [  88   93   84]]
[[4851344 6833884 3969792]
 [6833884 9627674 5591612]
 [3969792 5591612 3248656]]


The diagonal elements represent the dot product of each car vector with itself.

The other elements represent dot products between different cars.

However, engine capacity has much larger numerical values than speed and temperature. Therefore, it will dominate these dot products. Feature scaling should normally be applied before interpreting this matrix as a meaningful similarity measurement.

## 9. Calculating the Vector Magnitude

The Euclidean norm measures the magnitude or length of a vector.

For a vector:

$$x=[x_1,x_2,x_3]$$

its Euclidean norm is:

$$\lVert x\rVert
=
\sqrt{x_1^2+x_2^2+x_3^2}$$
`np.linalg.norm(car1)` calculates the overall numerical magnitude of the first car vector.

In [11]:
sensor_energy = np.linalg.norm(car1)
print(sensor_energy)



2202.5766728992658


The calculated value is the mathematical magnitude of the vector, not physical sensor energy.

Because the features use different units and scales, engine capacity has the largest effect on the result. Feature scaling would be required before using the norm to compare car samples fairly.

## Exploring Matrix Properties

The following square matrix is used to explore several linear algebra operations:

- The **determinant** checks whether the matrix is invertible.
- The **inverse** reverses the transformation performed by the matrix.
- The **eigenvalues and eigenvectors** describe the main directions and scaling effects of the matrix.

These operations are useful for understanding concepts used in machine learning, including matrix transformations and Principal Component Analysis.

In [12]:
A = np.array([[2 , 5],[3 , 7]])
print(A)
print(np.linalg.det(A))
print(np.linalg.inv(A))
print(np.linalg.eig(A))

[[2 5]
 [3 7]]
-1.0000000000000018
[[-7.  5.]
 [ 3. -2.]]
EigResult(eigenvalues=array([-0.11+0.j,  9.11+0.j]), eigenvectors=array([[-0.92+0.j, -0.58+0.j],
       [ 0.39+0.j, -0.82+0.j]]))


The determinant of matrix `A` is `-1`. Since it is not zero, the matrix is invertible.

`np.linalg.inv(A)` returns the inverse matrix, which satisfies:

$$A A^{-1} = I$$

`np.linalg.eig(A)` returns two results:

1. The eigenvalues of the matrix.
2. A matrix whose columns are the corresponding eigenvectors.

The eigenvalues show how strongly the matrix scales each eigenvector, while the eigenvectors represent directions that remain on the same line after the matrix transformation.

## 11. Creating a Shape Mismatch Error

Matrix multiplication requires the inner dimensions to match.

The car matrix has the shape:

$$cars:(3,3)$$

A valid weight vector must therefore contain exactly three elements.

The following incorrect weight vector contains four elements:

$$wrong\_weights:(4,)$$

The attempted multiplication becomes:

$$(3,3)@(4,)$$

This operation is invalid because the number of columns in the car matrix is 3, while the weight vector contains 4 elements.

In [13]:
wrong_weights = np.array([0.44, 0.11, 0.32, 0.002])

try:
    print(cars @ wrong_weights)
except ValueError as error:
    print("Shape mismatch error:")
    print(error)

Shape mismatch error:
matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 4 is different from 3)


### Why the Error Occurred

The error occurred because each car contains three features, but the incorrect weight vector contains four weights.

A model must have exactly one weight for every input feature.

The problem can be fixed by removing the extra value and creating a weight vector with three elements:

$$correct\_weights:(3,)$$

The corrected multiplication is:

$$(3,3)@(3,)\rightarrow(3,)$$

This produces one prediction for each car.

## 12. Generating Additional Car Samples

Five additional car samples are generated to test batch prediction on a larger dataset.

`np.random.seed(100)` makes the generated values reproducible. Running the notebook again will therefore generate the same samples.

The generated ranges represent:

- Engine capacity: from 1000 to 4999
- Speed: from 20 to 139
- Engine temperature: from 70 to 109

The resulting matrix contains five samples and three features:

$$random\_cars:(5,3)$$

In [14]:

np.random.seed(100)
random_cars = np.random.randint(low=[1000 , 20 , 70] , high=[5000 , 140 , 110] , size=(5,3))
print(random_cars)
print(random_cars @ weights)

[[2544   44   73]
 [4943  107   85]
 [2930  114  104]
 [1053   86  104]
 [2550   54   94]]
[44.23 85.33 86.59 66.96 53.45]


## Conclusion

In this notebook, I represented individual data samples as vectors and combined them into a dataset matrix. I used the dot product to calculate a prediction for one sample and verified the result using NumPy.

Matrix multiplication was then used to calculate predictions for multiple samples in one operation. I also examined matrix shapes and deliberately created a shape mismatch error to understand why the number of features must match the number of model weights.

Additional linear algebra operations, including transpose, vector magnitude, determinant, inverse, eigenvalues, and eigenvectors, were explored to connect basic matrix concepts with their applications in machine learning.